ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs

```diff
@@ -294,8 +303,30 @@ addBlockSync cdb@CDB {..} BlockToAdd { blockToAdd = b, .. } = do
         void $ chainSelectionForFutureBlocks cdb blockCache
         chainSelectionForBlock cdb blockCache hdr
 
+    -- There is a large comment in the body of
+    -- 'Ouroboros.Consensus.NodeKernel.initBlockFetchConsensusInterface' about
+    -- why this value should always be a 'Right'.
+    let whenItWasForged =
```

Comment:
This split out remains a TODO.

Also, I stubbornly labored to avoid a `MonadTime` constraint on this function. The `SupportsNode` constraint is similarly warty: this function doesn't need that constraint except to emit this event. That seems like poor separation of concerns (as evidence eg by this PR needing the `TestBlock` instance...).

If we paid the price to complicate the `Tracer` event types, then we could have a top-level type-changing tracer transformer that allowed for the `Tracer` here to know how to do this translation instead of this function itself knowing how. For downstream users, they'd just see an extra type parameter on the tracer events that Consensus emits, but it'd have a fix value by the time they care about it.

If we were to do this, then the `MonadTime` issue could be handled in the same way. Maybe it's preferable in the long-term to the current solution on this PR.

Reply:
I went through the PR once more, and I approved it as it looks good to me. My only concern is the additional 13 lines of code in the `addBlockSync` function that are there only for the sake of logging the `DoneAddingBlock` event. I wonder if this pollutes the logic of this function, or whether, on the contrary, describing how the forging time of a block can be calculated is pertinent to `addBlockSync`. Maybe we can talk about this in a call.